Mathlib/Analysis/Convex/Cone/Pointed.lean

```diff
@@ -0,0 +1,260 @@
+  · rw [← hzero, zero_smul]
+    exact S.zero_mem
+  · exact @ConvexCone.smul_mem 𝕜 E _ _ _ S _ _ hpos hx
+
+instance hasSmul : SMul { c : 𝕜 // 0 ≤ c } S where
+  smul := fun ⟨c, hc⟩ ⟨x, hx⟩ => ⟨c • x, S.smul_mem hc hx⟩
+
+instance hasNsmul : SMul ℕ S where
+  smul := fun n x => (n : { c : 𝕜 // 0 ≤ c }) • x
+
+@[simp]
+protected theorem coe_smul (x : S) (n : { c : 𝕜 // 0 ≤ c }) : n • x = n • (x : E) :=
+  rfl
+
+@[simp]
+protected theorem nsmul_eq_smul_cast (x : S) (n : ℕ) :
+    n • (x : E) = (n : { c : 𝕜 // 0 ≤ c }) • (x : E) :=
+  nsmul_eq_smul_cast _ _ _
+
+@[simp]
+theorem coe_nsmul (x : S) (n : ℕ) : (n • x : E) = n • (x : E) := by
+  -- simp_rw [PointedCone.coe_smul, PointedCone.nsmul_eq_smul_cast]; rfl
+  simp_rw [PointedCone.nsmul_eq_smul_cast]; rfl
+
+theorem add_mem ⦃x⦄ (hx : x ∈ S) ⦃y⦄ (hy : y ∈ S) : x + y ∈ S :=
+  S.add_mem' hx hy
+
+instance instAddMemClass : AddMemClass (PointedCone 𝕜 E) E where
+  add_mem ha hb := add_mem ha hb
+
+instance instAddCommMonoid : AddCommMonoid S :=
+  Function.Injective.addCommMonoid (Subtype.val : S → E) Subtype.coe_injective
+    rfl (by aesop) coe_nsmul
+    -- Note: linter says `coe_add` is a syntactic tautology
+
+/-- `addMonoidHom` structure on the inclusion map of a pointed cone inside the ambient space. -/
+def subtype.addMonoidHom : S →+ E where
+  toFun := Subtype.val
+  map_zero' := rfl
+  map_add' := by simp
+
+@[simp]
+theorem coeSubtype.addMonoidHom : (subtype.addMonoidHom : S → E) = Subtype.val := rfl
+
+instance instModule : Module { c : 𝕜 // 0 ≤ c } S := by
+  apply Function.Injective.module ({ c : 𝕜 // 0 ≤ c }) subtype.addMonoidHom
+  simp only [coeSubtype.addMonoidHom, Subtype.coe_injective]
+  simp -- a single `simp` does not work!
+
+/-- `linearMap` structure on the inclusion map of a pointed cone inside the ambient space. -/
+def subtype.linearMap : S →ₗ[{ c : 𝕜 // 0 ≤ c }] E where
+  toFun := Subtype.val
+  map_add' := by simp
+  map_smul' := by simp
+
+/-- A pointed cone is a submodule of the ambient when the scalars are restricted to be
+non-negative. -/
+def toSubmodule (S : PointedCone 𝕜 E) : Submodule { c : 𝕜 // 0 ≤ c } E where
+  carrier := S
+  add_mem' := fun hx hy => S.add_mem hx hy
+  zero_mem' := S.zero_mem
+  smul_mem' := fun ⟨c, hc⟩ x hx => by
+    simp_rw [SetLike.mem_coe]
+    cases' eq_or_lt_of_le hc with hzero hpos
+    · convert S.zero_mem
+      simpa [← hzero] using smul_eq_zero_of_left rfl x
+    · apply ConvexCone.smul_mem
+      convert hpos
+      exact hx
+
+/-- Any submodule (over non-negative scalars) of the ambient space is a pointed cone. -/
+def ofSubmodule (M : Submodule { c : 𝕜 // 0 ≤ c } E) : (PointedCone 𝕜 E) where
+  carrier := M
+  smul_mem' := fun c hc _ hx => M.smul_mem ⟨c, le_of_lt hc⟩ hx
+  add_mem' := fun _ hx _ hy => M.add_mem hx hy
+  zero_mem' := M.zero_mem
+
+/-- The equivalence between pointed cones and submodules. -/
+def toSubmoduleEquiv : (PointedCone 𝕜 E) ≃ (Submodule { c : 𝕜 // 0 ≤ c } E) where
```

Comment:
This makes me wonder: Can't you take this as the definition of `PointedCone` and reconstruct the API from there?

Reply:
I like this better as well. I'll redefine `PointedCone` and clean up the code.